File: src/code-samples/deepagents/models-runtime-configurable.py
"""Models: select a model at runtime with middleware."""

# :snippet-start: models-runtime-configurable-py
from dataclasses import dataclass
from typing import Callable

from langchain.agents.middleware import ModelRequest, ModelResponse, wrap_model_call
from langchain.chat_models import init_chat_model
from deepagents import create_deep_agent


@dataclass
class Context:
    model: str


@wrap_model_call
def configurable_model(
    request: ModelRequest,
    handler: Callable[[ModelRequest], ModelResponse],
) -> ModelResponse:
    model_name = request.runtime.context.model
    model = init_chat_model(model_name)
    return handler(request.override(model=model))


# KEEP MODEL
agent = create_deep_agent(
    model="google_genai:gemini-3.5-flash",
    middleware=[configurable_model],
    context_schema=Context,
)

# Invoke with the user's model selection
result = agent.invoke(
    {"messages": [{"role": "user", "content": "Hello!"}]},
    # KEEP MODEL
    context=Context(model="openai:gpt-5.5"),
)
# :snippet-end:

File: src/code-samples/deepagents/models-runtime-configurable.ts
// :snippet-start: models-runtime-configurable-js
import { createMiddleware, initChatModel } from "langchain";
import { createDeepAgent } from "deepagents";
import * as z from "zod";

const contextSchema = z.object({
  model: z.string(),
});

const configurableModel = createMiddleware({
  name: "ConfigurableModel",
  wrapModelCall: async (request, handler) => {
    const modelName = request.runtime.context.model;
    const model = await initChatModel(modelName);
    return handler({ ...request, model });
  },
});

// KEEP MODEL
const agent = await createDeepAgent({
  model: "google-genai:gemini-3.5-flash",
  middleware: [configurableModel],
  contextSchema,
});

// Invoke with the user's model selection
const result = await agent.invoke(
  { messages: [{ role: "user", content: "Hello!" }] },
  // KEEP MODEL
  { context: { model: "openai:gpt-5.5" } },
);
// :snippet-end:
